core/src/main/java/org/apache/james/core/MaybeSender.java

```diff
@@ -0,0 +1,91 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.core;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import com.google.common.collect.ImmutableList;
+
+public class MaybeSender {
```

Comment:
I'd like to see a `map` method.

Reply:
A `map` method that will transform this `MaybeSender` into an `Optional<Truc>` looks weird to me.

The namings that fits the best would be `<T> Optional<T> asOptional(Function<MailAddress, T> transformation)` but that looks overkill.

I can do it but I would like more feedback from the team.

Reply:
> you could have a real map method returning a MaybeSender.

We could but I'm not sure of the added value: we do not have `MailAddress -> MailAddress` transformations. Thus it would be unused... 

IMO `MaybeSender` is based on a monad but is not a monad.

It is a specific class that answers specific needs.

A proposal that would make more sense regarding callers would be:

```
Optional<String> getLocalPart();
Optional<Domain> getDomain();
```

But number of calls would be limited...